golang/cosmos/x/swingset/keeper/keeper.go

```diff
@@ -68,10 +69,49 @@ func NewKeeper(
 		accountKeeper:    accountKeeper,
 		bankKeeper:       bankKeeper,
 		feeCollectorName: feeCollectorName,
-		CallToController: callToController,
+		callToController: callToController,
 	}
 }
 
+// PushAction appends an action to the controller's action queue.  This queue is
+// kept in the kvstore so that changes to it are properly reverted if the
+// kvstore as rolled back.  By the time the block manager runs, it can commit
+// its SwingSet transactions without fear of side-effecting the world with
+// intermediate transaction state.
+func (k Keeper) PushAction(ctx sdk.Context, obj interface{}) error {
+	bz, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	tailStr := k.GetStorage(ctx, "actionQueue.tail")
+	var tail uint64
```

Comment:
The way `actionQueue.tail` is implicitly initialized by this code is subtle and worthy of a comment. Please also somewhere describe the full queue describing which side is responsible for what, and reference that description from the producer and the consumer.